index.js

```diff
@@ -82,6 +87,77 @@ const sendResponse = ( req, settings, fn ) => {
+	if ( expectStreamMode ) {
+		// Streaming responses is trickier than you might expect, with many footguns:
+		// • req.buffer(false): no version of superagent implements this when running in the browser
+		// • req.parse() or superagent.parse[]: only gets called when the response ends (see above)
+		// • req.on("progress"): doesn’t seem to work... at all
+		// • req.responseType(anything): makes superagent skip parse functions (see above)
+		// • req.xhr.responseType="blob": XHR only exposes partial responses in "" or "text" modes
+		// • req.xhr: only available after you call req.end()
+
+		// Expose partial responses.
+		// <https://xhr.spec.whatwg.org/#the-response-attribute>
+		req.xhr.responseType = 'text';
+
+		// Force the response to be treated as UTF-8. This is easier than the x-user-defined trick
+		// that would otherwise be needed to access the raw binary response body.
+		// <https://xhr.spec.whatwg.org/#final-charset>
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#31-serialization>
+		// <https://stackoverflow.com/a/33042003>
+		req.xhr.overrideMimeType( 'application/x-ndjson; charset=utf-8' );
+
+		// Find response chunks that end in a newline (possibly preceded by a carriage return), then
+		// for each chunk except the last, parse it as JSON and pass that to onStreamRecord.
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#31-serialization>
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#32-parsing>
+		// <https://stackoverflow.com/a/38440028>
+		let lastLine = null;
+		let start = 0;
+
+		req.xhr.addEventListener( 'progress', ( { target } ) => {
+			// Don’t use ProgressEvent#loaded in this algorithm. It measures progress in octets,
+			// while we’re working with text that has already been decoded from UTF-8 into a string
+			// that can only be indexed in UTF-16 code units. Reconciling this difference is not
+			// worth the effort, and might even be impossible if there were encoding errors.
+			let stop;
+			while ( true ) {
+				stop = target.response.indexOf( '\n', start );
+
+				if ( stop < 0 ) {
+					// Leave start untouched for the next progress event, waiting for the newline
+					// that indicates we’ve finished receiving a full line.
+					break;
+				}
+
+				// Note: not ignoring empty lines.
+				// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#32-parsing>
+				if ( lastLine !== null ) {
+					const record = JSON.parse( lastLine.trim() );
+					debug( 'stream mode: record=%o', record );
+					onStreamRecord( record );
+				}
+
+				lastLine = target.response.slice( start, stop );
+
+				// Make subsequent searches start *after* the newline.
+				start = stop + 1;
+			}
+		} );
+
+		// Parse the last response chunk as above, but pass it to the higher layers as The Response.
+		// Note: not ignoring empty lines.
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#32-parsing>
+		req.parse( () => JSON.parse( lastLine.trim() ) );
```

Comment:
you’ve understood that correctly!

> * onStreamRecord is fired with line N only after a full line N+1 arrives, so that we always guarantee that the last line will be in the final response, not in a streamrecord event.

this was true in 6d486a7f3f83ff13c999db24c84a48b44cfb7022, but in d385f273040368ae75e257fbd98b50bb3dce0003 i’ve tweaked the onStreamRecord line format to provide the same guarantees while eliminating that delay.

> Is it always guaranteed that we get a `progress` event even with the last part of the response?

i hadn’t considered that, but after consulting the XHR spec, yes. when we [#send](https://xhr.spec.whatwg.org/#the-send()-method), the [procedure](https://xhr.spec.whatwg.org/#handle-response-end-of-body) for when the end of the response has been received always includes firing a `progress` event (along with `load` and `loadend` events). i will document this.

> If the final response doesn't end with a newline, but simply with a closing `}` of the JSON, the code will fail to parse it. Is that fine?

yes. my understanding of [the ndjson spec](https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#31-serialization) is that senders must newline-terminate every line, but receivers aren’t required to accept otherwise.

Reply:
done in 936fd9e000518cb353f7836f309847f703abc515